### PR TITLE
Fix MutableOutputWidgets, and other bugs

### PR DESCRIPTION
**Pull request by gselzer**
This PR, motivated by #96, makes a couple fixes:
* Reacts to imagej/imagej-ops#647
* Re-includes the napari-imagej-specific scyjava `Converter`s, as the work done in #95 is not called by default
* Improves handling of optional parameters.

This PR cannot be merged until imagej/imagej-ops#647 is merged (and released), as the fixes depend on that work.

### CODE REVIEW

File: src/napari_imagej/types/widget_mappings.py [hunk 1]
```diff
@@ -36,8 +36,13 @@ def preferred_widget_for(
...
     """
-    if type_hint == "napari.layers.Image" and item.isInput() and item.isOutput():
-        return "napari_imagej.widgets.parameter_widgets.MutableOutputWidget"
+    if item.isInput() and item.isOutput():
+        if get_origin(type_hint) is Union and type(None) in get_args(type_hint):
+            type_hint = get_args(type_hint)[0]
+            if isinstance(type_hint, ForwardRef):
+                type_hint = type_hint.__forward_arg__
+        if type_hint == "napari.layers.Image":
+            return "napari_imagej.widgets.parameter_widgets.MutableOutputWidget"
```
Comment (gselzer) [inline src/napari_imagej/types/widget_mappings.py:45]: I feel like this is *super* confusing. Can we at least do it within a separate function/

File: src/napari_imagej/utilities/_module_utils.py [hunk 2]
```diff
@@ -163,11 +145,11 @@ def _filter_unresolved_inputs(
...
-    # Delegate optional output construction to the module
-    # We will leave those unresolved
-    unresolved = list(
-        filter(lambda i: not (i.isOutput() and not i.isRequired()), unresolved)
-    )
+    # # Delegate optional output construction to the module
+    # # We will leave those unresolved
+    # unresolved = list(
+    #     filter(lambda i: not (i.isOutput() and not i.isRequired()), unresolved)
+    # )
```
Comment (gselzer) [inline src/napari_imagej/utilities/_module_utils.py:152]: Remove this code

File: src/napari_imagej/utilities/_module_utils.py [hunk 3]
```diff
@@ -334,10 +313,11 @@ def _pure_module_outputs(
 
     outputs = module.getOutputs()
     for output_entry in outputs.entrySet():
-        # Ignore outputs that are also required inputs
+        # Ignore outputs that were provided by the user
         output_name = ij().py.from_java(output_entry.getKey())
         if module.getInfo().getInput(output_name) in user_inputs:
-            continue
+            if module.getInput(output_name):
+                continue
```
Comment (gselzer) [inline src/napari_imagej/utilities/_module_utils.py:320]: Might want to explain what's going on here better via comment

File: src/napari_imagej/utilities/_module_utils.py [hunk 4]
```diff
@@ -422,7 +402,7 @@ def _add_napari_metadata(
     # Without this, magicgui doesn't pick up on the types.
     type_hints = {str(i.getName()): python_type_of(i) for i in unresolved_inputs}
 
-    type_hints["return"] = _widget_return_type(info, unresolved_inputs)
+    type_hints["return"] = List[Layer]
```
Comment (gselzer) [inline src/napari_imagej/utilities/_module_utils.py:405]: I'd rather not hardcode in the type, here - can we make use of the function signature?

File: tests/types/test_converters.py [hunk 5]
```diff
@@ -755,3 +755,10 @@ def test_OutOfBoundsFactory_conversion(ij):
     assert isinstance(
         ij.py.to_java(OutOfBoundsFactory.PERIODIC), jc.OutOfBoundsPeriodicFactory
     )
+
+
+# -- Enum(like)s -- #
```
Comment (gselzer) [inline tests/types/test_converters.py:760]: This title should be images

File: tests/utilities/test_module_utils.py [hunk 6]
```diff
@@ -358,8 +359,8 @@ def func(*inputs):
 
     assert len(sig_params) == len(expected_names) + len(napari_param_map)
 
-    # assert return annotation is None
-    assert sig.return_annotation is None
+    # assert return annotation is stc
```
Comment (gselzer) [inline tests/utilities/test_module_utils.py:362]: typo `stc`

File: src/napari_imagej/utilities/_module_utils.py (hunk 2 above)
Comment (gselzer) [reply src/napari_imagej/utilities/_module_utils.py]: Removed in 003d620

File: src/napari_imagej/utilities/_module_utils.py (hunk 3 above)
Comment (gselzer) [reply src/napari_imagej/utilities/_module_utils.py]: Fixed with 3475c57

File: src/napari_imagej/types/widget_mappings.py (hunk 1 above)
Comment (gselzer) [reply src/napari_imagej/types/widget_mappings.py]: Cleaned up in 0bb96a3

**Comment by ctrueden** [comment]
The imagej-ops PR was merged, and released as imagej-ops 0.49.0.

File: tests/utilities/test_module_utils.py (hunk 6 above)
Comment (gselzer) [reply tests/utilities/test_module_utils.py]: Fixed and squashed into the correct commit.

File: tests/types/test_converters.py (hunk 5 above)
Comment (gselzer) [reply tests/types/test_converters.py]: Fixed and squashed into the correct commit

File: src/napari_imagej/utilities/_module_utils.py (hunk 4 above)
Comment (gselzer) [reply src/napari_imagej/utilities/_module_utils.py]: Cleaned up in 42e0463

File: src/napari_imagej/types/type_conversions.py [hunk 7]
```diff
@@ -54,6 +54,8 @@ def python_type_of(module_item: "jc.ModuleItem"):
     ):
         converted = converter(module_item)
         if converted is not None:
+            if not module_item.isRequired():
+                return Optional[converted]
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/types/type_conversions.py]: Now doing it inside checkers in a847088

File: src/napari_imagej/java.py [hunk 8]
```diff
@@ -58,8 +58,14 @@ def _imagej_init():
     )
     config.add_option(f"-Dimagej2.dir={os.getcwd()}")  # TEMP
     config.endpoints.append("io.scif:scifio:0.43.1")
+    config.endpoints.append("net.imagej:imagej-ops:0.49.0")
```
Comment (gselzer) [inline src/napari_imagej/java.py:61]: Should probably make notes to remove these when possible

Comment (gselzer) [reply src/napari_imagej/java.py:61]: And maybe issues as well

**Review by gselzer — COMMENTED** [review]
Super close!

File: src/napari_imagej/java.py (hunk 8 above)
Comment (gselzer) [reply src/napari_imagej/java.py:62]: Made #105, and removed the scifio pin. Also left a comment in.

**Pull request merged**